### DISCUSSION TITLE
Writing C# interface for spglib

### DISCUSSION THREAD

**Discussion by husakm** [spglib/spglib#399] [General]
Hi
I am not familiar with github mechanism. I had encoded a C# interface + sample code for spglib (porting just a few useful functions, not everything).  How can I share this code with other people interested in this topic ? E-mail me , please at husakm@vscht.cz.

Michal Husak
UCT Prague

**Comment by LecrisUT** [comment]
C# is compilable within CMake, so we can integrate it directly. More importantly, is there a common packaging system for C#? I remember there is Nuget, but is that popular outside the gaming community?

Other than that, are there conventions to distinguish C and C# headers like the `.hpp` for C++?

**Comment by husakm** [comment]
Dear Cristian Le

>C# is compilable within CMake, so we can integrate it directly. More 
>importantly, is there a common packaging system for C#? I remember 
>there is Nuget, but is that popular outside the gaming community?

I make the code as Microsoft Visual Studio project ... I have no 
experiences with CMake + the code depends on Microsoft libraries for 
user interaction + C#/C data exchange ...
If you want to make my sample accessible, I suggest to put simple in 
some sub directory with the Windows specific .sln project setup ....

For your curiosity, the import looks like followoing:

```csharp
using System;
using System.Runtime.InteropServices;           // DllImport
using System.Text; // StringBuilder

static class SPGLIB
     {

         [DllImport("symspg.dll")]
         public static extern int spg_get_international(StringBuilder symbol,
                                                double[,] lattice,
                                                double[,] position,
                                                int[] types,
                                                int num_atom,
                                                double symprec);
```

.....

>Other than that, are there conventions to distinguish C and C# 
>headers like the .hpp for C++?

There are no header files in C# = they are not necessary.
The available exported function list is automatically extracted from 
the added source code
or automatically extracted from the binary .dll file included to the 
projects ...

If you are interested I can download somewhere (pleas give me 
instructions) the current results
of my porting work (as Microsoft Visual Studio 2017 .sln project) ...

The code use spglib binary generated from .sln botained by CMake 
procesing of the Makefile ...
The spglib.h file need modification - add export statments to be able 
to generated dynamic version of the .dll.

Michal Husak
UCT Prague
associate professor

**Reply by LecrisUT** [reply]
> If you want to make my sample accessible, I suggest to put simple in
some sub directory with the Windows specific .sln project setup ....

Sure, we have `/examples` folder for that.

> For your curiosity, the import looks like followoing:
> 
> ```csharp
>          [DllImport("symspg.dll")]
>          public static extern int spg_get_international(StringBuilder symbol,
>                                                 double[,] lattice,
>                                                 double[,] position,
>                                                 int[] types,
>                                                 int num_atom,
>                                                 double symprec);
> ```

That is runtime import similar to `dlopen` in C/C++, or `cimport` in python. That is fine for an end-user project, and prototyping, but for having upstream C# support, we should make these imported "statically" (have the C# library link directly to `libsysmspg.dll`). We can do the latter directly via CMake and provide a `libspglib_csharp.dll` which would have interface like:
```csharp
using spglib;
using System;

class UserApplication {
  static void Main(string[] args) {
    double lattice[3][3] = {
        {3.111, -1.5555, 0},
        {0, 2.6942050311733885, 0},
        {0, 0, 4.988}
    };
    double position[4][3] = {
        {1.0 / 3, 2.0 / 3, 0.0},
        {2.0 / 3, 1.0 / 3, 0.5},
        {1.0 / 3, 2.0 / 3, 0.6181},
        {2.0 / 3, 1.0 / 3, 0.1181},
    };
    int types[4] = {1, 1, 2, 2};
    spglib.Dataset ds = Dataset(lattice, position, types);
    Console.WriteLine($"International symbol: {ds.international_symbol} ({ds.spacegroup_number})");
  }
}
```

> There are no header files in C# = they are not necessary.

Ok, that will make compatibility a bit easier. To make things public/private, it is sufficient to specify that in the source I guess?

> The `spglib.h` file need modification - add export statments to be able to generated dynamic version of the .dll.

We can do that. Currently we rely on:
```cmake
set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
It should export all symbols, so I am surprised that it did not work as is. But it would be good to move to scoped export, so I will gladly review the changes for that.

---

Currently we do not have windows developers, so it would be great if we can get some user feedback on such implementations. Overall there are a few things that we would need some help with:
- What is the native packaging system for C#, and how do you interact with it. E.g. how do you include a dependency on a `BLAS` library (not blas itself, but the equivalent C# library) so that it is accessible via `using blas;`
- How do you export these to end-user. My understanding is that `.sln` is meant for the end-project, but can it also be used to export the C# interface?
- Are there any well-known projects/communities that we can reference/contact for in packaging?
- What is the CI, unit-testing framework?

Also do you have any matrix/discord to discuss in more real-time? Would be nice to discuss the details of interface and get your initial example in the repo's `/examples`

**Comment by husakm** [comment]
Dear Christian

>That is runtime import similar to dlopen in C/C++, or cimport in 
>python. That is fine for an end-user project, and prototyping, but 
>for having upstream C# support, we should make these imported 
>"statically" (have the C# library link directly to libsysmspg.dll). 
>We can do the latter directly via CMake and provide a 
>libspglib_csharp.dll which would have interface like:

I am not sure  why you can not import the function from the dynamic .dll ...
We do this in our code from OpenGL and it works ...
For upgrade, it is sufficient to change the spglib.dll and all works ...

>using spglib;
>using System;
>
>class UserApplication {
>   static void Main(string[] args) {
>     double lattice[3][3] = {
>         {3.111, -1.5555, 0},
>         {0, 2.6942050311733885, 0},
>         {0, 0, 4.988}
>     };
>     double position[4][3] = {
>         {1.0 / 3, 2.0 / 3, 0.0},
>         {2.0 / 3, 1.0 / 3, 0.5},
>         {1.0 / 3, 2.0 / 3, 0.6181},
>         {2.0 / 3, 1.0 / 3, 0.1181},
>     };
>     int types[4] = {1, 1, 2, 2};
>     spglib.Dataset ds = Dataset(lattice, position, types);
>     Console.WriteLine($"International symbol: 
> {ds.international_symbol} ({ds.spacegroup_number})");
>   }
>}

So you already have some C# sample code ?
Can I have a look ?

>We can do that. Currently we rely on:
>set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)It should export all 
>symbols, so I am surprised that it did not work as is. But it would 
>be good to move to scoped export, so I will gladly review the changes for that.

All was working under Windows for static spglib build ...
For the dynamic version (witch I use from C#) it was required to add 
manually to spglib.h sequence like:

#ifdef DYNAMIC
#define EXPORT __declspec(dllexport)
#else
#define EXPORT
#endif

and than:
EXPORT int spg_get_major_version(void);

The CMAKE did not modified the header file ....

>
>----------
>Currently we do not have windows developers, so it would be great if 
>we can get some user feedback on such implementations. Overall there 
>are a few things that we would need some help with:

I am both a Linux and Windows developer. Under Linux I work on 
modification of QM programs (CASTEP, Quantum Espresso) witch use spglib ...
Under Windows I write in C# crystallographic software for the QM 
programs preparations ...

>    * What is the native packaging system for C#, and how do you 
> interact with it. E.g. how do you include a dependency on a BLAS 
> library (not blas itself, but the equivalent C# library) so that it 
> is accessible via using blas;
I just distribute the sample import .cs source code + usage 
examples.  You add it to your project and you get the functionality.
We access all external libraries (OpenGL, win32api, drivers of 
devices, blas, fft) by the mechanism above = C#  Interop service.
using System.Runtime.InteropServices;           // DllImport

>    * How do you export these to end-user. My understanding is that 
> .sln is meant for the end-project, but can it also be used to 
> export the C# interface?
As mentioned, we give to the user witch want to include the spglib to 
hsi project simple 3 things:

1) The dynamic spglib.dll version generated from your C code
2) SPGLIB.cs file witch contains the translation of the structures 
and fucntions from C code to C#. This is a class witchc an be called 
from other C# codes.
Including this file to the projects exports the interfaces to the 
whole project ...
3) Sample C# code witch use 1) + 2)

For final product distribution it is required to place the C# 
executabel in the smae directory as spglib.dll ...

>    * Are there any well-known projects/communities that we can 
> reference/contact for in packaging?
I do not know. We do not pack the code ...
>    * What is the CI, unit-testing framework?
I am not familiar with CI. I test the code simply by writing C# 
program using the code ...

>Also do you have any matrix/discord to discuss in more real-time? 
>Would be nice to discuss the details of interface and get your 
>initial example in the repo's /examples

Unfortunately I do not use Discored nor Matric ...
I can send you simple sample immediately ....
Unfortunately it will not go through E-mail (security reasons) ...
Can you , please, give me instruction where I can upload ?

Michal

**Reply by LecrisUT** [reply]
> I am not sure  why you can not import the function from the dynamic .dll. We do this in our code from OpenGL and it works. For upgrade, it is sufficient to change the spglib.dll and all works.

Various issues:
- It does not check the `SOVERSION` of the library to make sure they are compatible
- It is inefficient and hard to debug when the loading of the dll fails
- Prone to memory-leak because it does not have proper contructor/destructor

> So you already have some C# sample code? Can I have a look?

Unfortunately no, but I can write up the interface shown above. The issue is getting an environment to test locally since I haven't written anything in C# for a few years, and that may take a week or so.

> ```c
> #ifdef DYNAMIC
> #define EXPORT __declspec(dllexport)
> #else
> #define EXPORT
> #endif
> ```

Ok, this one I can handle. Didn't think that any language would actually check for those when importing the header.

> I do not know. We do not pack the code

I am afraid that this is a very important aspect that I want to strongly discourage in scientific coding. It makes reusing the code from project-to-project virtually impossible. This is why I am maintaining the packaging systems. I can research this, but it might take 2-3 weeks to get up to speed.

> Can you , please, give me instruction where I can upload?

Can you upload them to https://gist.github.com/ and provide a link to that. Just add each relevant files there and their content and `create public gist`.

**Comment by husakm** [comment]
Hi

>I am afraid that this is a very important aspect that I want to 
>strongly discourage in scientific coding. It makes reusing the code 
>from project-to-project virtually impossible. This is why I am 
>maintaining the packaging systems. I can research this, but it might 
>take 2-3 weeks to get up to speed.

I was elaborating with this when I was importing OpenGL glu and Win32 
api futures in my code ...
There exist C# .dll making this directly = you put the C# .dll in the 
directory where you make C# development
and the compiler extracts the interfaces to the class from the .dll .
In reality it does not work OK ...  If there is any problem with the 
.dll you can not handle this on your side ...
So I prefer to include the interface to the C .dll in C# as C# source 
code directly withtoiu any addtional C# .dll layer ...

>Can you upload them to 
><https://gist.github.com/>https://gist.github.com/ and provide a 
>link to that. Just add each relevant files there and their content 
>and create public gist.

I just want to upload somewhere the complete Visual Studio Project 
including the executable and .dll, so direct test will be possible ...
Github does not accept the .zip files .
Some other idea for binary file sharing ?

Michal

**Reply by LecrisUT** [reply]
> If there is any problem with the .dll you can not handle this on your side

But that's precisely why it would be provided by us, so that we catch and fix any dll issues and we make sure we are linking with the appropriate versions.

But in the meantime, I am not against providing instructions for end-user linkage.

> including the executable and .dll

Don't, 90% of the time, it would not be executable on our side since we do not have windows and the linkage will be broken. The project files is sufficient. I can manage myself around building the project from scratch. We also **must** catch any OS dependencies issues, before adding it to `/examples`.

**Comment by LecrisUT** [comment]
Note to self:
- https://github.com/actions/setup-dotnet#github-package-registry-gpr
- `try ... catch (EntryPointNotFoundException)` otherwise call `SetDllDirectory` to local and try again. Hopefully it works. https://stackoverflow.com/questions/8836093/how-can-i-specify-a-dllimport-path-at-runtime#8861895
- How to deal with different names for linux/windows in `dllimport`?
- `P/invoke` vs `dllimport`?

**Comment by husakm** [comment]
I worked some more time on the C# interface
Is there any chance to change following functional calls in spglib:
A)
`SpglibSpacegroupType spg_get_spacegroup_type(int hall_number);`

To something like:
B)
` SpglibError spg_get_spacegroup_type_ext(SpglibSpacegroupType *spglibtype, int hall_number);`

?
Or at least to have the functions duplicate in both versions for backward compatibility.
The already used variant A) is a marshal/implementation nightmary.
The version B) can be interfaced from C# in a relatively clean way by commands like:

```
public struct SpglibSpacegroupType
        {
            public int number;
            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 11)]
            public string international_short;
            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 20)]
            public string international_full;
            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 32)]
            public string international;
            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 7)]
            public string schoenflies;
            public int hall_number;
            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 17)]
            public string hall_symbol;
            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 6)]
            public string choice;
            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 6)]
            public string pointgroup_international;
            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 4)]
            public string pointgroup_schoenflies;
            public int arithmetic_crystal_class_number;
            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 7)]
            public string arithmetic_crystal_class_symbol;
        }

    [DllImport("symspg.dll")]
    public static extern SpglibError spg_get_spacegroup_type_ext(out SpglibSpacegroupType spglibtype,
                                                             int hall_number);
```

and later used in main user C# code simple and elegant as:


```
SPGLIB.SpglibError result;
SPGLIB.SpglibSpacegroupType spgtype = new SPGLIB.SpglibSpacegroupType();
result = SPGLIB.spg_get_spacegroup_type_ext(out spgtype, 446);
```

**Reply by LecrisUT** [reply]
I am thinking of introducing a C++ interface. How is the C# <-> C++ integration? I hope it can help with `A`.

Regarding the interface I would prefer C# to take ownership of the data, e.g. look at the Fortran API. It copies the data and then deconstructs the C object. The reason is that I don't trust the interface to maintain ownership.

**Comment by husakm** [comment]
There is no way to help with A)
Generally the A) is considered as sub-optimal code design, because the compilers can do anything with the data.
If you have heard about the COM technology - functions returns only error codes, never results directly ...
For B) the data are under control of the C# application.
I have already extended spglib by the B option for selected functions + make tests ...
C# <-> C++ integration is identical as C# <-> C integration. No difference ...
I will make a Windows sample code + project  and put it somewhere for downlod so you will exactly see what I am doing ...

**Comment by husakm** [comment]
I had put the test project on Google Drive:
https://drive.google.com/drive/folders/1Qwfo8wlWQSMhPOKvUCOCyJNJck3Kv3Vf
It contains the C# spglib interface, binary example + windows binary symspg.dll  for experiments. For one fucntion a test B) implementation is given + a demo why A) does not work ...
The file name is TestSPGLIB_08.zip

**Reply by husakm** [reply]
Sorry the link shodul be:
https://drive.google.com/file/d/15ON85GizQQ5FyfdjZ8vvZ-y0SpucrrqF/view

**Reply by LecrisUT** [reply]
Can you make it public? I don't have a google account

**Reply by husakm** [reply]
This link shodul be public:
https://drive.google.com/file/d/15ON85GizQQ5FyfdjZ8vvZ-y0SpucrrqF/view?usp=sharing
I cna not chec this on my side - I am aot looged to my account ....

**Reply by LecrisUT** [reply]
Thanks, it works now, I'll give a quick look, but won't review it until I finish up a few dangling PRs

**Comment by husakm** [comment]
I had set the public argument to the file, so this should work already:
https://drive.google.com/file/d/15ON85GizQQ5FyfdjZ8vvZ-y0SpucrrqF/view?usp=sharing

**Comment by husakm** [comment]
I had created a more advanced example of spglib usage in C#. Free download for a project and Windows binary in VC 2017 format:

https://drive.google.com/file/d/1GY5C2kWFdjHj5Ttg4KpZDKxoW4CSTYxi/view?usp=sharing

It is impossible to write an interface from C# to spglib without adding new export functions and structure to C spglib (due to issues with data Marshaling I am not able to handle). So I added to the C code of spglib a few new structures+functions named _ext with C# friendly interfaces ...
If there will be an interest, I can re-create in C# all spglib functions + published modified spglib C code + C# SDK with the C# SPGLIB.cs implementing an interface with C# friendly Spglib structure equivalents ...
As it is visible from my source code I prefer 100% code safety = I declare all data structures on C# side and transfer to C only pointers to this structures. C copy to this pre-made structures the results from the C code part. So the memory management is done on C# side in a safe way.
Pleas let me know what do you thing + suggestions for improvement.
I am able to provide a C# SDK based on source code sharing, but I ma not able to make some sophisticated distribution packages (nuget ?)
I belive the spglib will be primary used by developers so sharing source code can be a more flexible way than sharing a C# binary .dll with no way of modification possible ...
I have ready a sample code, visualizing in C# structures read form .CIF files by OpenGL as well ... I had encoded an interface to SPGLIb in this code so space group checking is possible ... If anybody is interested I can share source ...

**Reply by LecrisUT** [reply]
Sorry I would need a bit more time until I can review this, maybe 2-3 weeks. Do ping me if it falls off my radar

**Reply by LecrisUT** [reply]
@husakm Quick point though, can you convert the project to `.NET Core 8` (or anything `.NET Core`)? I cannot build a `.NET Framework` on my Fedora, but I did manage to setup Rider for `.NET Core 8`

**Reply by husakm** [reply]
The code can run only on Windows and should be compiled by Microsoft Visual Studio ... I have no way to do anything working on Fedora ....

**Reply by LecrisUT** [reply]
https://learn.microsoft.com/en-us/visualstudio/get-started/csharp/tutorial-console?view=vs-2022

When you select the target framework there should be a `.Net 8.0` option. Here are screenshots from my Rider IDE, but the same applies to MVS. I don't have any Windows environments to give you the equivalent MVS.

![Screenshot_20240214_192606](https://github.com/spglib/spglib/assets/25577083/a3601563-d971-49c8-9176-92d8b6d1e98b)
![Screenshot_20240214_192629](https://github.com/spglib/spglib/assets/25577083/7d4b8618-6198-42fa-8854-2a2a82e929bc)

**Reply by LecrisUT** [reply]
At least I got something to work after I changed the `symspg.dll` to `symspg`:
![Screenshot_20240214_194423](https://github.com/spglib/spglib/assets/25577083/24b6a611-331e-403d-9a63-daff947412ea)

So far it seems I can replicate the project and I can look more closely at the design. I'll come back to it in maybe 2 weeks

**Comment by husakm** [comment]
Hi


The code use Microsoft .NET for visualizastion + spglib modified binary for Windows.

You can work with it only on Window by the help of Microsft Visual Studio ....


Michal


________________________________

**Comment by husakm** [comment]
The curent version of my demo use .NET for showing window + interface  ....
I will make a super simple demo with command line only interface
witch will be probably possible to port more easy on Linux ....
I will share the modified spglib C source as welll
Give me 2 weeks ....
The one thing I am not sure is the comunicaton betwen C# and
.dll on Linux .... I am not sure the compilers will alighn correctly
memory for the data sharing ...

Btw. I was today testing spglib on abput 200 structures from Cabridge structure databde ...
It had discovered in 2 cases incorrect space group assigment ... Confirmed experimdntaly later ...
________________________________
